cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactStore.java

```diff
@@ -64,26 +66,58 @@
  * plugin classes and/or application classes. We may want to extend this to include other types of classes, such
  * as datasets.
  *
- * Every time an artifact is added, the artifact contents are stored at a location based on its id:
+ * Every time an artifact is added, the artifact contents are stored at a base location based on its id:
  * /namespaces/{namespace-id}/artifacts/{artifact-name}/{artifact-version}
+ * Several metadata writes are then performed.
  *
- * Several writes are also performed on the meta table, one for metadata about the artifact itself, and one
- * for each entity contained in the artifact.
+ * The first adds metadata about the artifact, with
+ * rowkey r:{namespace}:{artifact-name}, column {artifact-version}, and ArtifactDetail as the value
  *
- * Artifact metadata is stored with r:{namespace}:{artifact-name} as the rowkey, {artifact-version} as the column,
- * and ArtifactMeta as the value.
+ * TODO: (CDAP-2764) add this part when we have a better idea of what needs to be in AppClass.
+ * The second adds metadata about any Application Class contained in the artifact, with
+ * rowkey a:{namespace}:{classname}, column {artifact-name}:{artifact-version}, and AppClass as the value
  *
- * PluginClass metadata is stored with p:{namespace}:{plugin-type}:{plugin-name} as the rowkey,
- * {artifact-version}:{artifact-name} as the column, and PluginClass as the value
+ * The third adds metadata about any Plugin contained in the artifact, with
+ * rowkey p:{parent-namespace}:{parent-name}:{plugin-type}:{plugin-name},
+ * column {artifact-namespace}:{artifact-name}:{artifact-version},
+ * and PluginDetailCodec as the value
  *
- * TODO: (CDAP-2764) add this part when we have a better idea of what needs to be in AppClass.
- * AppClass metadata is stored with a:{namespace}:{app-name} as the rowkey,
- * {artifact-version}:{artifact-name} as the column, and AppClass as the value
+ * For example, suppose we add a system artifact etlbatch-3.1.0, which contains an ETLBatch application class.
+ * The meta table will look like:
+ *
+ * rowkey                            columns
+ * a:system:ETLBatch                 etlbatch:3.1.0 -> {AppData}
+ * r:system:etlbatch                 3.1.0 -> {ArtifactData}
+ *
+ * After that, a system artifact etlbatch-lib-3.1.0 is added, which extends etlbatch and contains
+ * stream sink and table sink plugins. The meta table will look like:
+ *
+ * rowkey                            columns
+ * a:system:ETLBatch                 etlbatch:3.1.0 -> {AppData}
+ * p:system:etlbatch:sink:stream     system:etlbatch-lib:3.1.0 -> {PluginData}
+ * p:system:etlbatch:sink:table      system:etlbatch-lib:3.1.0 -> {PluginData}
```

Comment:
Just to confirm, PluginSelector can be used to not only choose among different versions of the artifacts with same name, but also among different artifact names. So a plugin named table of type sink, can be part of artifacts with different names and versions?